File: frontend/src/types/case.ts
export interface Case {
  id: number;
  fir_number: string;
  crime_head_id: number | null;
  crime_subhead_id: number | null;
  district_id: number;
  unit_id: number | null;
  investigating_officer_id: number | null;
  case_category_id: number | null;
  case_status_id: number | null;
  gravity_id: number | null;
  occurrence_date: string | null;
  fir_date: string | null;
  latitude: number | null;
  longitude: number | null;
  brief_facts: string | null;
}

export interface CaseCreate {
  fir_number: string;
  crime_head_id: number;
  crime_subhead_id: number;
  district_id: number;
  unit_id: number;
  investigating_officer_id: number;
  case_category_id: number;
  case_status_id?: number;
  gravity_id: number;
  occurrence_date?: string | null;
  fir_date?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  brief_facts?: string | null;
}

export interface CaseUpdate {
  crime_head_id?: number;
  crime_subhead_id?: number;
  investigating_officer_id?: number;
  case_status_id?: number;
  occurrence_date?: string | null;
  brief_facts?: string | null;
  latitude?: number | null;
  longitude?: number | null;
}

export interface AccusedByCaseResponse {
  case_id: number;
  fir_number: string;
  accused: {
    accused_id: number;
    accused_name: string;
    status: string | null;
  }[];
}

export interface AccusedSearchResult {
  accused_id: number;
  accused_name: string;
  case_id: number;
  status: string | null;
}

export type CaseStatusName =
  | "Under Investigation"
  | "Chargesheet Filed"
  | "Closed"
  | "Pending Trial"
  | "Convicted"
  | "Acquitted";

File: backend/app/schemas/case.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel


class CaseCreate(BaseModel):
    fir_number: str
    crime_head_id: int
    crime_subhead_id: int
    district_id: int
    unit_id: int
    investigating_officer_id: int
    case_category_id: int
    case_status_id: int = 1
    gravity_id: int
    occurrence_date: Optional[datetime] = None
    fir_date: Optional[datetime] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    brief_facts: Optional[str] = None


class CaseUpdate(BaseModel):
    crime_head_id: Optional[int] = None
    crime_subhead_id: Optional[int] = None
    investigating_officer_id: Optional[int] = None
    case_status_id: Optional[int] = None
    occurrence_date: Optional[datetime] = None
    brief_facts: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None


class CaseOut(BaseModel):
    id: int
    fir_number: str
    crime_head_id: Optional[int] = None
    crime_subhead_id: Optional[int] = None
    district_id: int
    unit_id: Optional[int] = None
    investigating_officer_id: Optional[int] = None
    case_category_id: Optional[int] = None
    case_status_id: Optional[int] = None
    gravity_id: Optional[int] = None
    occurrence_date: Optional[datetime] = None
    fir_date: Optional[datetime] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    brief_facts: Optional[str] = None

    model_config = {"from_attributes": True}
